### PR TITLE
require that the user and github token owner match

### PR DESCRIPTION
**Pull request by dhellmann**

### CODE REVIEW

**Comment by mgedmin** [comment]
Oof, the [cog failure on git master](https://github.com/mgedmin/ghcloneall/runs/7315400083?check_suite_focus=true) not being recognized as failure of the overall build means that I was right to suspect `continue-on-error: true` for linter jobs.

Should be fixed on git master.

File: ghcloneall.py [hunk 1]
```diff
@@ -441,6 +450,7 @@ def list_repos(self, user=None, organization=None, pattern=None,
                         '?sort=full_name').format(
                             owner)
         elif user and not organization:
+            self._verify_user_token(user)
```
Comment (mgedmin) [inline ghcloneall.py]: I think this should be moved inside the `if include_private and self.has_auth_token:`

File: tests.py [hunk 2]
```diff
@@ -46,6 +48,15 @@ def __init__(self):
...
         self.responses.update(responses)
 
+    def set_user(self, user):
+        if user is None:
+            if self.user_endpoint in self.responses:
+                del self.responses[self.user_endpoint]
+        else:
+            self.responses[self.user_endpoint] = MockResponse(
+                json={'login': user},
+            )
```
Comment (mgedmin) [inline tests.py:58]: I'd like to make the mock a bit smarter and return a response only when the request contains an Authorization header, failing with a 401 otherwise.

I think that would have caught the use case where you're now calling _verify_user_token() even if the user hasn't provided a token.

File: tests.py [hunk 3]
```diff
@@ -1389,23 +1401,23 @@ def test_main_run_error_handling_with_private_token(monkeypatch, capsys):
...
 def test_main_run_error_handling_no_private_token(monkeypatch, capsys):
     monkeypatch.setattr(sys, 'argv', [
-        'ghcloneall', '--user', 'mgedmin',
+        'ghcloneall', '--user', 'test_user',
     ])
     with pytest.raises(SystemExit) as ctx:
         ghcloneall.main()
     assert str(ctx.value) == (
-        'Failed to fetch https://api.github.com/users/mgedmin/repos'
+        'Failed to fetch https://api.github.com/users/test_user/repos'
```
Comment (mgedmin) [inline tests.py]: It would likely make the diff a bit smaller with these unrelated username changes in tests.

File: tests.py [hunk 4]
```diff
@@ -1446,14 +1458,26 @@ def test_main_run_with_token(monkeypatch, mock_requests_get, capsys):
     )
 
 
+def test_main_run_with_mismatched_token(monkeypatch, mock_requests_get,
+                                        capsys):
+    monkeypatch.setattr(sys, 'argv', [
+        'ghcloneall', '--user', 'test_user', '--concurrency=1',
+        '--github-token', 'fake-token',
+    ])
+    mock_requests_get.set_user('some-other-user')
```
Comment (mgedmin) [inline tests.py:1470]: You know, having this test here I think we can avoid writing smart mock response classes by defaulting to .set_user(None) in the fixture, and setting it to something only in tests that actually provide a token.

(I apologize for the spam of multiple single comments, I don't know why I started doing that instead of a proper preview.  Too much GitLab at work?)

**Review by mgedmin — CHANGES_REQUESTED** [review]
Overall response: I think this breaks unauthenticated access again by doing the token check too soon, which will return a 401 from GitHub, which IIRC will raise an exception from the get_json_and_links() helper, aborting the process.

Details in single comments.

File: ghcloneall.py (hunk 1 above)
Comment (dhellmann) [reply ghcloneall.py]: D'oh, I messed that up when I refactored it to put it into a method.

File: tests.py (hunk 3 above)
Comment (dhellmann) [reply tests.py]: Yeah, I put them in a separate commit but maybe that work is not needed at all. I'll see what I get when I fix the other issue and remove the name changes.

File: ghcloneall.py (hunk 1 above)
Comment (dhellmann) [reply ghcloneall.py]: Fixed

File: tests.py (hunk 3 above)
Comment (dhellmann) [reply tests.py]: I removed that change and only set the user in a couple of the tests where it's needed.

File: tests.py (hunk 4 above)
Comment (dhellmann) [reply tests.py:1470]: Done

**Comment by dhellmann** [comment]
I've updated this PR to fix the location of the validation and remove the extra changes in the tests.

File: tests.py (hunk 2 above)
Comment (dhellmann) [reply tests.py:58]: So if the user is None, the response should be 401 instead of failing with an error because the mock wasn't completely configured?

Comment (mgedmin) [reply tests.py:58]: I've just checked: `get_json_and_links()` converts all 4xx errors to an exception, so I don't think the distinction between 401 and 404 really matters for these tests.

**Review by mgedmin — APPROVED** [review]
LGTM!

Thank you very much!
